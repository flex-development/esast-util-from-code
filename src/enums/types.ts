/**
 * @file Enums - types
 * @module esast-util-from-code/enums/types
 */

import type { EsastNode } from '@flex-development/esast'
import type { Type } from '@flex-development/unist-util-types'

/**
 * esast node types.
 *
 * @enum {Type<EsastNode>}
 */
enum types {
  ambientDeclaration = 'ambientDeclaration',
  argumentList = 'argumentList',
  arithmeticExpression = 'arithmeticExpression',
  arrayExpression = 'arrayExpression',
  arrayPattern = 'arrayPattern',
  arrayType = 'arrayType',
  arrowFunctionExpression = 'arrowFunctionExpression',
  asClause = 'asClause',
  asExpression = 'asExpression',
  assertionPredicate = 'assertionPredicate',
  assignmentExpression = 'assignmentExpression',
  assignmentPattern = 'assignmentPattern',
  awaitExpression = 'awaitExpression',
  bigint = 'bigint',
  bitwiseExpression = 'bitwiseExpression',
  blockStatement = 'blockStatement',
  boolean = 'boolean',
  breakStatement = 'breakStatement',
  callExpression = 'callExpression',
  callSignature = 'callSignature',
  catchClause = 'catchClause',
  classBody = 'classBody',
  classDeclaration = 'classDeclaration',
  classExpression = 'classExpression',
  classHeritage = 'classHeritage',
  comment = 'comment',
  computedExpression = 'computedExpression',
  conditionalExpression = 'conditionalExpression',
  conditionalType = 'conditionalType',
  constructSignature = 'constructSignature',
  constructorType = 'constructorType',
  continueStatement = 'continueStatement',
  debuggerStatement = 'debuggerStatement',
  decorator = 'decorator',
  decoratorList = 'decoratorList',
  directive = 'directive',
  doWhileStatement = 'doWhileStatement',
  elseStatement = 'elseStatement',
  emptyStatement = 'emptyStatement',
  enumBody = 'enumBody',
  enumDeclaration = 'enumDeclaration',
  enumMember = 'enumMember',
  equalityExpression = 'equalityExpression',
  exportDeclaration = 'exportDeclaration',
  exportSpecifier = 'exportSpecifier',
  exportSpecifiers = 'exportSpecifiers',
  expressionStatement = 'expressionStatement',
  extendsClause = 'extendsClause',
  fieldDefinition = 'fieldDefinition',
  finallyBlock = 'finallyBlock',
  forInStatement = 'forInStatement',
  forOfStatement = 'forOfStatement',
  forStatement = 'forStatement',
  fromClause = 'fromClause',
  functionDeclaration = 'functionDeclaration',
  functionExpression = 'functionExpression',
  functionSignature = 'functionSignature',
  functionType = 'functionType',
  genericType = 'genericType',
  identifier = 'identifier',
  ifStatement = 'ifStatement',
  implementsClause = 'implementsClause',
  importAssertion = 'importAssertion',
  importAttribute = 'importAttribute',
  importDeclaration = 'importDeclaration',
  importExpression = 'importExpression',
  importMeta = 'importMeta',
  importSpecifier = 'importSpecifier',
  importSpecifiers = 'importSpecifiers',
  indexSignature = 'indexSignature',
  indexedAccessType = 'indexedAccessType',
  inferType = 'inferType',
  interfaceBody = 'interfaceBody',
  interfaceDeclaration = 'interfaceDeclaration',
  intersectionType = 'intersectionType',
  labeledStatement = 'labeledStatement',
  logicalExpression = 'logicalExpression',
  mappedSignature = 'mappedSignature',
  mappedType = 'mappedType',
  memberExpression = 'memberExpression',
  methodDefinition = 'methodDefinition',
  methodSignature = 'methodSignature',
  modifier = 'modifier',
  modifierList = 'modifierList',
  moduleBody = 'moduleBody',
  moduleDeclaration = 'moduleDeclaration',
  moduleExpression = 'moduleExpression',
  namespaceDeclaration = 'namespaceDeclaration',
  namespaceExport = 'namespaceExport',
  namespaceImport = 'namespaceImport',
  newExpression = 'newExpression',
  nonNullExpression = 'nonNullExpression',
  nothing = 'nothing',
  null = 'null',
  number = 'number',
  objectExpression = 'objectExpression',
  objectLiteralType = 'objectLiteralType',
  objectPattern = 'objectPattern',
  parameter = 'parameter',
  parameterList = 'parameterList',
  parenthesizedExpression = 'parenthesizedExpression',
  parenthesizedType = 'parenthesizedType',
  property = 'property',
  propertySignature = 'propertySignature',
  regexp = 'regexp',
  relationalExpression = 'relationalExpression',
  restElement = 'restElement',
  restType = 'restType',
  returnStatement = 'returnStatement',
  root = 'root',
  satisfiesClause = 'satisfiesClause',
  satisfiesExpression = 'satisfiesExpression',
  sequenceExpression = 'sequenceExpression',
  spreadElement = 'spreadElement',
  staticBlock = 'staticBlock',
  staticBlockBody = 'staticBlockBody',
  string = 'string',
  super = 'super',
  switchBody = 'switchBody',
  switchCase = 'switchCase',
  switchStatement = 'switchStatement',
  taggedTemplateExpression = 'taggedTemplateExpression',
  templateElement = 'templateElement',
  templateLiteral = 'templateLiteral',
  templatePlaceholder = 'templatePlaceholder',
  this = 'this',
  throwStatement = 'throwStatement',
  tryStatement = 'tryStatement',
  tupleElement = 'tupleElement',
  tupleType = 'tupleType',
  typeAliasDeclaration = 'typeAliasDeclaration',
  typeAnnotation = 'typeAnnotation',
  typeArgumentList = 'typeArgumentList',
  typeAssertion = 'typeAssertion',
  typeAssertionExpression = 'typeAssertionExpression',
  typeKeyword = 'typeKeyword',
  typeParameter = 'typeParameter',
  typeParameterList = 'typeParameterList',
  typePredicate = 'typePredicate',
  unaryExpression = 'unaryExpression',
  unaryTypeExpression = 'unaryTypeExpression',
  undefined = 'undefined',
  unionType = 'unionType',
  updateExpression = 'updateExpression',
  variableDeclaration = 'variableDeclaration',
  variableDeclarator = 'variableDeclarator',
  whileStatement = 'whileStatement',
  withStatement = 'withStatement',
  yieldExpression = 'yieldExpression'
}

export default types
