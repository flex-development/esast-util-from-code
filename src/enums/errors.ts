/**
 * @file Enums - errors
 * @module esast-util-from-code/enums/errors
 */

/**
 * Error messages.
 *
 * @enum {string}
 */
enum errors {
  comment = 'Unterminated comment',
  string = 'Unterminated string constant'
}

export default errors
