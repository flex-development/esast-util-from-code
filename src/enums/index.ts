/**
 * @file Entry Point - Enums
 * @module esast-util-from-code/enums
 */

export { default as errors } from './errors'
export { default as tt } from './tt'
