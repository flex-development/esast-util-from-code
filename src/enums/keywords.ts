/**
 * @file Enums - keywords
 * @module esast-util-from-code/enums/keywords
 */

/**
 * Possible keywords.
 *
 * @enum {Lowercase<string>}
 */
enum keywords {
  abstract = 'abstract',
  accessor = 'accessor',
  any = 'any',
  as = 'as',
  assert = 'assert',
  asserts = 'asserts',
  async = 'async',
  await = 'await',
  bigint = 'bigint',
  boolean = 'boolean',
  break = 'break',
  case = 'case',
  catch = 'catch',
  class = 'class',
  const = 'const',
  constructor = 'constructor',
  continue = 'continue',
  debugger = 'debugger',
  declare = 'declare',
  default = 'default',
  delete = 'delete',
  do = 'do',
  else = 'else',
  enum = 'enum',
  export = 'export',
  extends = 'extends',
  false = 'false',
  finally = 'finally',
  for = 'for',
  from = 'from',
  function = 'function',
  get = 'get',
  if = 'if',
  implements = 'implements',
  import = 'import',
  in = 'in',
  infer = 'infer',
  instanceof = 'instanceof',
  interface = 'interface',
  keyof = 'keyof',
  let = 'let',
  meta = 'meta',
  module = 'module',
  namespace = 'namespace',
  never = 'never',
  new = 'new',
  null = 'null',
  number = 'number',
  object = 'object',
  of = 'of',
  out = 'out',
  override = 'override',
  private = 'private',
  protected = 'protected',
  public = 'public',
  readonly = 'readonly',
  return = 'return',
  satisfies = 'satisfies',
  set = 'set',
  static = 'static',
  string = 'string',
  super = 'super',
  switch = 'switch',
  symbol = 'symbol',
  this = 'this',
  throw = 'throw',
  true = 'true',
  try = 'try',
  type = 'type',
  typeof = 'typeof',
  undefined = 'undefined',
  unique = 'unique',
  unknown = 'unknown',
  using = 'using',
  var = 'var',
  void = 'void',
  while = 'while',
  with = 'with',
  yield = 'yield'
}

export default keywords
