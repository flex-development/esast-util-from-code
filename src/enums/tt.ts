/**
 * @file Enums - tt
 * @module esast-util-from-code/enums/tt
 */

import type { TokenType } from '@flex-development/vfile-lexer'

/**
 * Token kinds.
 *
 * @see {@linkcode TokenType}
 *
 * @enum {TokenType}
 */
enum tt {
  bigint = 'bigint',
  comment = 'comment',
  eof = 'eof',
  keyid = 'keyid',
  number = 'number',
  punctuator = 'punctuator',
  sof = 'sof',
  string = 'string',
  whitespace = 'whitespace'
}

export default tt
