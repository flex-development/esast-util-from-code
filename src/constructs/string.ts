/**
 * @file Constructs - string
 * @module esast-util-from-code/constructs/string
 */

import { errors, tt } from '#src/enums'
import {
  codes,
  resolveSlice,
  type Code,
  type Construct,
  type Effects,
  type Guard,
  type Resolver,
  type State,
  type TokenizeContext
} from '@flex-development/vfile-lexer'
import { ok as assert } from 'devlop'

/**
 * Opening punctuator.
 *
 * @var {typeof codes.apostrophe | typeof codes.quotation} opener
 */
let opener: typeof codes.apostrophe | typeof codes.quotation

/**
 * Check if the previous character `code` can come before this construct.
 *
 * @see {@linkcode Code}
 * @see {@linkcode TokenizeContext}
 *
 * @this {TokenizeContext}
 *
 * @param {Code} code - Previous character code
 * @return {boolean} `true` if `code` allowed before construct
 */
function previous(this: TokenizeContext, code: Code): boolean {
  return code !== codes.backslash
}

/**
 * Check if the current character `code` can start this construct.
 *
 * @see {@linkcode Code}
 * @see {@linkcode TokenizeContext}
 *
 * @this {TokenizeContext}
 *
 * @param {Code} code - Current character code
 * @return {boolean} `true` if `code` can start construct
 */
function test(this: TokenizeContext, code: Code): boolean {
  if (code === codes.apostrophe) return opener = code, true
  if (code === codes.quotation) return opener = code, true
  return false
}

/**
 * String literal construct.
 *
 * @see https://developer.mozilla.org/docs/Web/JavaScript/Reference/Lexical_grammar#string_literals
 *
 * @const {Construct} string
 */
const string: Construct = {
  /**
   * Construct name.
   */
  name: tt.string,

  /**
   * Check if the previous character code can come before this construct.
   *
   * @see {@linkcode Guard}
   */
  previous,

  /**
   * Resolve the events parsed by `tokenize`.
   *
   * @see {@linkcode Resolver}
   */
  resolve: resolveSlice,

  /**
   * Check if the current character code can start this construct.
   *
   * @see {@linkcode Guard}
   */
  test,

  /**
   * Set up a state machine to handle character codes streaming in.
   *
   * @see {@linkcode Effects}
   * @see {@linkcode State}
   * @see {@linkcode TokenizeContext}
   *
   * @this {TokenizeContext}
   *
   * @param {Effects} effects - Context object to transition state machine
   * @param {State} ok - Successful tokenization state
   * @return {State} Initial state
   */
  tokenize(this: TokenizeContext, effects: Effects, ok: State): State {
    /**
     * Tokenize context.
     *
     * @const {TokenizeContext} self
     */
    const self: TokenizeContext = this

    return string

    /**
     * Finish string literal tokenization.
     *
     * @param {Code} code - Current character code
     * @return {State | undefined} Next state
     */
    function finish(code: Code): State | undefined {
      if (code === opener && previous.call(self, self.previous)) {
        effects.consume(code)
        effects.exit(tt.string)
        return ok
      }

      /**
       * Unterminated string check.
       *
       * @var {boolean} terminated
       */
      let unterminated: boolean = false

      // unterminated string check
      switch (true) {
        case code === codes.eof:
        case self.isLineEnding(code):
          unterminated = true
          break
        case opener === codes.apostrophe && code === codes.quotation:
        case opener === codes.quotation && code === codes.apostrophe:
          if (previous.call(self, self.previous)) unterminated = true
          break
        default:
          break
      }

      // raise on unterminated string
      if (unterminated) {
        self.raise(errors.string, self.now(), {
          actual: code === codes.eof ? code : self.serialize(code),
          expected: [self.serialize(opener)],
          ruleId: tt.string
        })

        effects.exit(tt.string)
        return ok(code)
      }

      effects.consume(code)
      return finish
    }

    /**
     * Start string literal tokenization.
     *
     * @param {Code} code - Current character code
     * @return {State} Next state
     */
    function string(code: Code): State {
      assert(code === opener, 'expected `opener`')
      return effects.enter(tt.string), effects.consume(code), finish
    }
  }
}

export default string
