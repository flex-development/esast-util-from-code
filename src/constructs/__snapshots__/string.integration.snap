// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`integration:constructs/string > double quote > should tokenize string (escaped opener) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 string "\\"\\\\\\"\\"" (1:1-1:5, 0-4)
│     whitespace: ""
└─2 eof null (1:5-1:5, 4-4)
      whitespace: ""
`;

exports[`integration:constructs/string > double quote > should tokenize string (unterminated, eof) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 string "\\"string" (1:1-1:8, 0-7)
│     whitespace: ""
└─2 eof null (1:8-1:8, 7-7)
      whitespace: ""
`;

exports[`integration:constructs/string > double quote > should tokenize string (unterminated, line ending) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 string "\\"string" (1:1-1:8, 0-7)
│     whitespace: ""
└─2 eof null (2:1-2:1, 8-8)
      whitespace: "\\n"
`;

exports[`integration:constructs/string > double quote > should tokenize string (unterminated, opener mismatch) 1`] = `
tokens[4]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 string "\\"string" (1:1-1:8, 0-7)
│     whitespace: ""
├─2 string "''" (1:8-1:10, 7-9)
│     whitespace: ""
└─3 eof null (1:10-1:10, 9-9)
      whitespace: ""
`;

exports[`integration:constructs/string > double quote > should tokenize string 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 string "\\"string\\"" (1:1-1:9, 0-8)
│     whitespace: ""
└─2 eof null (1:9-1:9, 8-8)
      whitespace: ""
`;

exports[`integration:constructs/string > single quote > should tokenize string (escaped opener) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 string "'\\\\''" (1:1-1:5, 0-4)
│     whitespace: ""
└─2 eof null (1:5-1:5, 4-4)
      whitespace: ""
`;

exports[`integration:constructs/string > single quote > should tokenize string (unterminated, eof) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 string "'string" (1:1-1:8, 0-7)
│     whitespace: ""
└─2 eof null (1:8-1:8, 7-7)
      whitespace: ""
`;

exports[`integration:constructs/string > single quote > should tokenize string (unterminated, line ending) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 string "'string" (1:1-1:8, 0-7)
│     whitespace: ""
└─2 eof null (2:1-2:1, 8-8)
      whitespace: "\\n"
`;

exports[`integration:constructs/string > single quote > should tokenize string (unterminated, opener mismatch) 1`] = `
tokens[4]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 string "'string" (1:1-1:8, 0-7)
│     whitespace: ""
├─2 string "\\"\\"" (1:8-1:10, 7-9)
│     whitespace: ""
└─3 eof null (1:10-1:10, 9-9)
      whitespace: ""
`;

exports[`integration:constructs/string > single quote > should tokenize string 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 string "'string'" (1:1-1:9, 0-8)
│     whitespace: ""
└─2 eof null (1:9-1:9, 8-8)
      whitespace: ""
`;
